lib/routes/utils/signin.js

```diff
@@ -0,0 +1,485 @@
+      }
+      return password.verifyHash()
+        .then(verifyHash => {
+          return db.checkPassword(accountRecord.uid, verifyHash)
+        })
+        .then(match => {
+          if (match) {
+            return match
+          }
+          return customs.flag(clientAddress, {
+            email: accountRecord.email,
+            errno: error.ERRNO.INCORRECT_PASSWORD
+          }).then(() => {
+            match
+          })
+        })
+    },
+
+    /**
+     * Check if the user is logging in with the correct email address
+     * for their account.  If not, throw an error that will help them
+     * find the right one.
+     */
+    checkEmailAddress(accountRecord, email, originalLoginEmail) {
+      // The `originalLoginEmail` param is specified in the login request if the user has
+      // changed their primary email address, and hence needed to use a different email for
+      // calculating authPW than they use for identifying their account.
+      if (originalLoginEmail) {
+        // They must have been originally trying to log in with the correct
+        // primary email address, not a secondary one.
+        if (originalLoginEmail.toLowerCase() !== accountRecord.primaryEmail.normalizedEmail) {
+          throw error.cannotLoginWithSecondaryEmail()
+        }
+
+        // If a user changes their primary email and then deletes the old primary
+        // email from their account, that email address is basically "bricked" -
+        // the user can't log in with it, and nobody can create a new account with it.
+        // The below causes us to return a special error when trying to sign in
+        // with such a "bricked" email, because otherwise the default error message
+        // would offer the user to create an account with that email, which they can't.
+        let usingValidEmail = false
+        accountRecord.emails.some((emailData) => {
+          if (emailData.normalizedEmail === originalLoginEmail.toLowerCase()) {
+            usingValidEmail = true
+            return true
+          }
+        })
+        if (! usingValidEmail) {
+          throw error.cannotLoginWithEmail()
+        }
+      } else if (email.toLowerCase() !== accountRecord.primaryEmail.normalizedEmail) {
+        throw error.cannotLoginWithSecondaryEmail()
+      }
+      return P.resolve(true)
+    },
+
+    /**
+     * Check if user is allowed a password-checking attempt, and if so then
+     * load their accountRecord.  These two operations are intertwined due
+     * to the "unblock codes" feature, which allows users to bypass customs
+     * checks and which requires us to load the account record for processing.
+     *
+     * Returns an object with the following information about the process:
+     *
+     *  {
+     *    accountRecord:     the user's account record loaded form the db
+     *    didSigninUnblock:  whether an unblock code was successfully used
+     *  }
+     */
+    checkCustomsAndLoadAccount(request, email) {
```

Comment:
I tried to unwind and comment the steps of this procedure a little more so they're easier to follow; in the old version of the code this logic was spread across a couple of different helper functions.